<?xml version="1.0"?>

<launch>
  <!-- Topics -->
  <arg name="attitude_state_topic" default="attitude_state" />
  <arg name="lander_reference_force_topic" default="lander_reference_force" />
  <arg name="actuator_commands_topic" default="actuator_commands" />
  <arg name="positional_state_topic" default="positional_state" />
  <arg name="reference_positional_state_topic" default="reference_positional_state" />
  <arg name="feedforward_force_topic" default="feedforward_force" />
  <arg name="lander_reference_force_topic" default="lander_reference_force" />
  <arg name="vicon_transform_topic" default="vicon/robot_1/robot_1" />

  <!-- Services -->
  <arg name="esc_enable_service" default="EnableESC" />
  <arg name="esc_disable_service" default="DisableESC" />
  <arg name="servo_enable_service" default="EnableServo" />
  <arg name="servo_disable_service" default="DisableServo" />
  <arg name="takeoff_service" default="Takeoff" />
  <arg name="land_service" default="Land" />

  <!-- Attitude Control Parameters -->
  <param name="AttitudeProportionalGain" default="" />
  <param name="AttitudeDerivativeGain" default="" />
  <param name="AttitudeYawDerivativeGain" default="" />
  <param name="GimbalDistance" default="" />
  <param name="MaxGimbalAngle" default="" />
  <param name="PropellerThrustCoefficient" default="" />
  <param name="PropellerMomentCoefficient" default="" />
  <param name="MaxThrustDifferential" default="" />
  <param name="AttitudeControlRate" default="" />

  <!-- Positional Control Parameters -->
  <param name="FeedbackGainMatrix" default="" />
  <param name="MaxPointingAngle" default="" />
  <param name="MaxThrust" default="" />
  <param name="MinThrust" default="" />
  <param name="PositionalControlRate" default="" />

  <!-- Propeller Parameters -->
  <param name="esc_channel_1" default="" />
  <param name="esc_channel_2" default="" />

  <!-- Servo Parameters -->
  <param name="servo_channel_1" default="" />
  <param name="servo_channel_2" default="" />

  <!-- Hover Parameters-->
  <param name="HoverHeight" default="" />
  <param name="HoverFeedforwardForce" default="" />

  <!-- Estimation Parameters-->
  <param name="ViconEstimatorRate" default="" />

  <include file="$(find erl_quadrotor_vicon)/launch/erl_vicon.launch" />

  <node name="vicon_estimator_node"
        pkg="lander_junior_ros"
        type="StateEstimateVicon.py"
        output="screen">
    <param name="topics/vicon_transform" value="$(arg vicon_transform_topic)" />
    <param name="topics/attitude_state" value="$(arg attitude_state_topic)" />
    <param name="topics/positional_state" value="$(arg positional_state_topic)" />
  </node>

  <node name="attitude_controller_node"
        pkg="lander_junior_ros"
        type="AttitudeControl.py"
        output="screen">
    <param name="topics/attitude_state" value="$(arg attitude_state_topic)" />
    <param name="topics/lander_reference_force" value="$(arg lander_reference_force_topic)" />
    <param name="topics/actuator_commands" value="$(arg actuator_commands_topic)" />
  </node>

  <node name="positional_controller_node"
        pkg="lander_junior_ros"
        type="PositionalControl.py"
        output="screen">
    <param name="topics/positional_state" value="$(arg positional_state_topic)" />
    <param name="topics/reference_positional_state" value="$(arg reference_positional_state_topic)" />
    <param name="topics/feedforward_force" value="$(arg feedforward_force_topic)" />
    <param name="topics/lander_reference_force" value="$(arg lander_reference_force_topic)" />
  </node>

  <node name="servo_node"
        pkg="lander_junior_ros"
        type="Servo"
        output="screen">
    <param name="topics/actuator_commands" value="$(arg actuator_commands_topic)" />
    <param name="services/EnableServo" value="$(arg servo_enable_service)" />
    <param name="services/DisableServo" value="$(arg servo_disable_service)" />
  </node>

  <node name="esc_node"
        pkg="lander_junior_ros"
        type="ESC"
        output="screen">
    <param name="topics/actuator_commands" value="$(arg actuator_commands_topic)" />
    <param name="services/EnableESC" value="$(arg esc_enable_service)" />
    <param name="services/DisableESC" value="$(arg esc_disable_service)" />
  </node>

  <node name="takeoff_server_node"
        pkg="lander_junior_ros"
        type="TakeoffServer.py"
        output="screen">
    <param name="topics/positional_state" value="$(arg positional_state_topic)" />
    <param name="topics/reference_positional_state" value="$(arg reference_positional_state_topic)" />
    <param name="topics/feedforward_force" value="$(arg feedforward_force_topic)" />
    <param name="services/EnableESC" value="$(arg esc_enable_service)" />
    <param name="services/DisableESC" value="$(arg esc_disable_service)" />
    <param name="services/EnableServo" value="$(arg servo_enable_service)" />
    <param name="services/DisableServo" value="$(arg servo_disable_service)" />
    <param name="services/Takeoff" value="$(arg takeoff_service)" />
    <param name="services/Land" value="$(arg land_service)" />
  </node>

</launch>




